File: sudoku-common/common-mybatis/src/main/java/com/zumin/sudoku/common/mybatis/utils/PageUtils.java
package com.zumin.sudoku.common.mybatis.utils;

import static java.util.stream.Collectors.toList;

import com.github.pagehelper.PageHelper;
import com.github.pagehelper.PageInfo;
import com.zumin.sudoku.common.core.utils.ServletUtils;
import com.zumin.sudoku.common.mybatis.enums.MybatisStatusCode;
import com.zumin.sudoku.common.mybatis.excpetion.PageException;
import com.zumin.sudoku.common.mybatis.page.Page;
import com.zumin.sudoku.common.mybatis.page.PageConvert;
import com.zumin.sudoku.common.mybatis.page.PageParam;
import java.util.List;
import java.util.function.Function;
import lombok.experimental.UtilityClass;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Component;

/**
 * 分页工具类
 */
@UtilityClass
public class PageUtils {

  /**
   * 最大分页条数
   */
  private final int MAX_PAGE_SIZE = 20;
  /**
   * 每页显示条数的参数名称
   */
  private final String PAGE_SIZE_NAME = "pageSize";
  /**
   * 默认查询的页数
   */
  private final int DEFAULT_PAGE = 1;
  /**
   * 默认每页个数
   */
  private final int DEFAULT_PAGE_SIZE = 5;
  /**
   * 当前查询页的参数名称
   */
  private final String PAGE_NAME = "page";

  /**
   * 获取分页数据
   *
   * @param pageParam 分页参数
   * @param <T>       源数据类型
   * @return 分页数据
   */
  public <T> Page<T> getPage(@NotNull PageParam<T> pageParam) {
    PageTemplateMethod<T> pageTemplateMethod = new PageTemplateMethod<>(pageParam);
    return pageTemplateMethod.getPage(PageInfo::of);
  }

  /**
   * 获取分页数据，并利用转换器转化为指定类型
   *
   * @param pageParam 分页参数
   * @param converter 源数据转为目标数据的转换器
   * @param <T>       源数据类型
   * @param <V>       目标数据类型
   * @return 分页数据
   */
  @SuppressWarnings("all")
  public <T, V> Page<V> getPage(@NotNull PageParam<T> pageParam, @NotNull Function<T, V> converter) {
    PageTemplateMethod<T> pageTemplateMethod = new PageTemplateMethod<>(pageParam);
    return (Page<V>) pageTemplateMethod.getPage(queryList -> {
      List<V> targetList = queryList.stream().map(converter).collect(toList());
      PageInfo pageInfo = PageInfo.of(queryList);
      pageInfo.setList(targetList);
      return pageInfo;
    });
  }

  /**
   * 获取分页详情对象的回调方法
   *
   * @param <T> 分页数据的类型
   */
  private interface GetPageInfoCallBack<T> {

    /**
     * 获取分页详情对象
     *
     * @param queryList 查询数据列表
     * @return 分页详情对象
     */
    PageInfo<T> getPageInfo(List<T> queryList);
  }

  /**
   * 获取分页数据的模板方法
   *
   * @param <T> 分页数据的类型
   */
  private static class PageTemplateMethod<T> {

    private final PageParam<T> pageParam;

    public PageTemplateMethod(PageParam<T> pageParam) {
      this.pageParam = pageParam;
    }

    /**
     * 获取分页数据
     *
     * @param callBack 获取分页详情对象的回调方法
     * @return 分页数据
     */
    @SuppressWarnings("all")
    public Page<T> getPage(GetPageInfoCallBack<T> callBack) {
      PageHelper.startPage(getPage(), getPageSize());
      List<T> queryList = pageParam.getQueryFunc().get();
      return PageConvert.INSTANCE.convert(callBack.getPageInfo(queryList));
    }

    /**
     * 获取查询页数
     *
     * @return 查询页数
     */
    private int getPage() {
      Integer page = pageParam.getPage();
      return page != null ? page : ServletUtils.getParameterToInt(PAGE_NAME, DEFAULT_PAGE);
    }

    /**
     * 获取每页个数
     *
     * @return 每页个数
     */
    private int getPageSize() {
      Integer pageSize = pageParam.getPageSize();
      if (pageSize == null) {
        pageSize = ServletUtils.getParameterToInt(PAGE_SIZE_NAME, DEFAULT_PAGE_SIZE);
      }
      if (pageSize <= 0 || pageSize > MAX_PAGE_SIZE) {
        throw new PageException(MybatisStatusCode.PAGE_SIZE_ILLEGAL);
      }
      return pageSize;
    }
  }
}

File: sudoku-common/common-mybatis/src/main/java/com/zumin/sudoku/common/mybatis/page/PageUtils.kt
package com.zumin.sudoku.common.mybatis.page

import com.github.pagehelper.PageHelper
import com.github.pagehelper.PageInfo
import com.zumin.sudoku.common.core.utils.getFromRequestToInt
import com.zumin.sudoku.common.mybatis.MybatisStatusCode

// 最大分页条数
private const val MAX_PAGE_SIZE = 20

// 每页显示条数的参数名称
private const val PAGE_SIZE_NAME = "pageSize"

// 默认查询的页数
private const val DEFAULT_PAGE = 1

// 默认每页个数
private const val DEFAULT_PAGE_SIZE = 5

// 当前查询页的参数名称
private const val PAGE_NAME = "page"

/**
 * 获取分页数据
 *
 * @param <T>       源数据类型
 * @return 分页数据
 */
fun <T> PageParam<T>.getPageData(): Page<T> {
  val pageTemplateMethod = PageTemplateMethod(this)
  return pageTemplateMethod.getPage { PageInfo.of(it) }
}

/**
 * 获取分页数据，并利用转换器转化为指定类型
 *
 * @param converter 源数据转为目标数据的转换器
 * @param <T>       源数据类型
 * @param <V>       目标数据类型
 * @return 分页数据
 */
fun <T, V> PageParam<T>.getPageData(converter: (T) -> V): Page<V> {
  return PageTemplateMethod(this).getPage { list ->
    val targetList = list.map(converter)
    PageInfo.of(list).apply { setList(targetList as List<T>) }
  } as Page<V>
}

/**
 * 获取分页数据的模板方法
 *
 * @param <T> 分页数据的类型
 */
private class PageTemplateMethod<T>(private val pageParam: PageParam<T>) {
  /**
   * 获取分页数据
   *
   * @param callBack 获取分页详情对象的回调方法
   * @return 分页数据
   */
  fun getPage(callBack: (List<T>) -> PageInfo<T>): Page<T> {
    PageHelper.startPage<Any>(page, pageSize)
    val queryList = pageParam.queryFunc.get()
    return callBack.invoke(queryList).toPage()
  }

  /**
   * 获取查询页数
   *
   * @return 查询页数
   */
  private val page = pageParam.page ?: PAGE_NAME.getFromRequestToInt(DEFAULT_PAGE)

  /**
   * 获取每页个数
   *
   * @return 每页个数
   */
  private val pageSize: Int
    get() {
      val pageSize = pageParam.pageSize ?: PAGE_SIZE_NAME.getFromRequestToInt(DEFAULT_PAGE_SIZE)
      if (pageSize <= 0 || pageSize > MAX_PAGE_SIZE) {
        throw PageException(MybatisStatusCode.PAGE_SIZE_ILLEGAL)
      }
      return pageSize
    }
}
